src/axbolt/delayedmessage/repository.go

```diff
@@ -0,0 +1,193 @@
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/ptypes"
+	"github.com/jmalloc/ax/src/ax"
+	"github.com/jmalloc/ax/src/ax/endpoint"
+	"github.com/jmalloc/ax/src/ax/marshaling"
+	"github.com/jmalloc/ax/src/ax/persistence"
+	"github.com/jmalloc/ax/src/axbolt/internal/boltutil"
+	boltpersistence "github.com/jmalloc/ax/src/axbolt/persistence"
+)
+
+// Repository is a Bolt-backed implementation of Ax's delayedmessage.Repository
+// interface.
+type Repository struct{}
+
+const (
+	// DelayedMessageBktName is the name of the Bolt root bucket where delayed
+	// messages are stored.
+	DelayedMessageBktName = "ax_delayed_message"
+
+	// BySendAtBktName is the name of a subbucket in DelayedMessageBktName where
+	// delayed messages are stored and indexed by SendAt field in message's
+	// envelope
+	BySendAtBktName = "by_send_at"
+
+	// ByIDBktName is the name of a subbucket in DelayedMessageBktName where
+	// delayed messages are stored and indexed by MessageID field in message's
+	// envelope
+	ByIDBktName = "by_id"
+)
+
+// LoadNextMessage loads the next that is scheduled to be sent.
+func (Repository) LoadNextMessage(
+	ctx context.Context,
+	ds persistence.DataStore,
+) (endpoint.OutboundEnvelope, bool, error) {
+	db := boltpersistence.ExtractDB(ds)
+	tx, err := db.Begin(false)
+	if err != nil {
+		return endpoint.OutboundEnvelope{}, false, err
+	}
+	defer tx.Rollback()
+
+	bkt := boltutil.GetBkt(
+		tx,
+		DelayedMessageBktName,
+		BySendAtBktName,
+	)
+	if bkt == nil {
+		return endpoint.OutboundEnvelope{}, false, nil
+	}
+
+	k, pb := bkt.Cursor().First()
+	if k != nil && pb == nil {
+		return endpoint.OutboundEnvelope{}, false, nil
+	}
+
+	m, err := parseDelayedMessage(pb)
+	if err != nil {
+		return endpoint.OutboundEnvelope{}, false, err
+	}
+	return m, true, nil
+}
+
+// SaveMessage saves a message to be sent at a later time.
+// If does NOT return an error if the message already exists in the repository.
+func (Repository) SaveMessage(
+	ctx context.Context,
+	ptx persistence.Tx,
+	env endpoint.OutboundEnvelope,
+) error {
+	var err error
+	tx := boltpersistence.ExtractTx(ptx)
+	// return nil if this is a duplicate entry
+	if m := boltutil.Get(
+		tx,
+		env.MessageID.Get(),
+		DelayedMessageBktName,
+		ByIDBktName,
+	); m != nil {
+		return nil
+	}
+
+	m := &DelayedMessage{
+		Id:                  env.MessageID.Get(),
+		CausationId:         env.CausationID.Get(),
+		CorrelationId:       env.CorrelationID.Get(),
+		CreatedAt:           marshaling.MarshalTime(env.CreatedAt),
+		SendAt:              marshaling.MarshalTime(env.SendAt),
+		Operation:           int32(env.Operation),
+		DestinationEndpoint: env.DestinationEndpoint,
+	}
+	if m.Message, err = ptypes.MarshalAny(env.Message); err != nil {
+		return err
+	}
+
+	if err = boltutil.Put(
+		tx,
+		env.MessageID.Get(),
+		[]byte(m.SendAt),
+		DelayedMessageBktName,
+		ByIDBktName,
+	); err != nil {
+		return err
+	}
+
+	return boltutil.PutProto(
+		tx,
+		m.SendAt,
```

Comment:
What would happen if 2 messages had the same `SendAt` time (however unlikely)?

Reply:
My first though was to use Message ID as a suffix but then it will most likely affect message sorting by key bytes which is highly critical for Delayed Message package. Can we suffix the key with generated bucket sequence?